ebpf/state.go

```diff
@@ -275,6 +277,102 @@ func (ns *networkState) mergeConnections(id string, active map[string]*Connectio
 	return conns
 }
 
+func commonPorts(m map[uint16]int) map[uint16]struct{} {
+	ports := make(map[uint16]struct{})
+	for p, count := range m {
+		if count > 25 { // TODO: Make configurable
+			ports[p] = struct{}{}
+		}
+	}
+	return ports
+}
+
+func aggregateConnections(conns []ConnectionStats) []ConnectionStats {
+	aggrConns := make([]ConnectionStats, 0)
+
+	sort.Slice(conns, func(i, j int) bool {
+		if conns[i].Pid < conns[j].Pid { // Sort first by PID
+			return true
+		} else if conns[i].Pid == conns[j].Pid { // Then by destination IP
+			if comp := strings.Compare(conns[i].Dest, conns[j].Dest); comp == -1 {
+				return true
+			} else if comp == 0 { // Then by source IP
+				return strings.Compare(conns[i].Source, conns[j].Source) == -1
+			}
+		}
+		return false
+	})
+
+	// Go over PID sub-ranges
+	applyFuncOnRange(conns,
+		func(a, b *ConnectionStats) bool {
+			return a.Pid == b.Pid
+		},
+		func(pidRange []ConnectionStats) {
+			// Find common ports for this PID range
+			staticSource, staticDest := map[uint16]int{}, map[uint16]int{}
+			for _, c := range pidRange {
+				staticDest[c.DPort]++
+				staticSource[c.SPort]++
+			}
+			commonSource, commonDest := commonPorts(staticSource), commonPorts(staticDest)
+
+			// Go over IP sub-ranges and aggregate
+			applyFuncOnRange(pidRange,
+				func(a, b *ConnectionStats) bool {
+					return a.Source == b.Source && a.Dest == b.Dest
```

Comment:
do you want to rollup if `a.source == b.source` OR `a.dest == b.dest` ?

Reply:
I would say no, because then we'll merge connections from different IP <> IP pairs

Reply:
This condition creates a slice of all connections that share the same source and dest IPs

Reply:
https://goplay.space/#hiCzzkbEJQ0 here's an example of `applyFuncOnRange()` with integers

Reply:
Ah I thought source was a full address, not just IP

Reply:
That said, with this approach we'd be folding

```
a:10 -> b:200
a:11 -> b:200
a:35 -> b:200

a:40 -> b:400
a:41 -> b:400
...
a:64 -> b:400
```

Into one connection instead of two. I could see this being a problem for redis where it's common to have several of the same process each listening on different ports.

Reply:
...if you have 25 connections open to each instance which I suppose is not terribly likely

Reply:
Sorry I don't quite understand. We still also fold on common static ports, but only over an (IP, IP) range.

In the above case, we'll create two connections:

```
a:* -> b:200
a:* -> b:400
```

But iff the total number of connections to *:200 (and similarly for *:400) on this process are >= 25